design/node-agent-concurrency.md

```diff
@@ -0,0 +1,77 @@
+
+## Goals
+
+- Define the behaviors of concurrent VGDP instances in node-agent
+- Create a mechanism for users to specify the concurrent number of VGDP per node
+
+## Non-Goals
+- VGDP instances from different nodes always run in concurrent since in most common cases the resources are isolated. For special cases that some resources are shared across nodes, there is no support at present
+- In practice, restores run in prioritized scenarios, e.g., disarster recovery. However, the current design doesn't consider this difference, a VGDP instance for a restore is blocked if it reaches to the limit of the concurrency, even though the ones block it are for backups. If users do meet some problems here, they should consider to stop the backups first
+
+## Solution
+
+### Global concurrent number
+We add a node-agent server parameter ```data-path-concurrent-num``` to accept a concurrent number that will be applied to all nodes if the per-node number is not specified.  
+The number starts from 1 which means there is no concurrency, only one instance of VGDP is allowed. There is no roof limit.    
+If the node-agent server parameter is not specified, a hard-coded default value will be used, the value is set to 1.   
+
+### Per-node concurrent number
+We allow users to specify different concurrent number per node, for example, users can allow 3 concurrent instances in Node-1, 2 instances in Node-2 and 1 instance in Node-3. This is for below considerations:
+- The resources may be different among nodes. Then users could specify smaller concurrent number for nodes with less resources while larger number for the ones with more resources
+- Help users to isolate critical environments. Users may run some critical workloads in some specified nodes, since VGDP instances may take large resource consumption, users may only want to run them in the other nodes or run less number of instances in the nodes with critical workloads
+
+The range of Per-node concurrent number is the same with Global concurrent number.  
+Per-node concurrent number is preferable to Global concurrent number, so it will overwrite the Global concurrent number for that node.  
+Per-node concurrent number is implemented in a separate configMap named ```node-agent-configs```. This configMap is not created by Velero, users should create it manually on demand. The configMap should be created before node-agent server starts, otherwise, node-agent server should be restarted in order to make the effect.  
+
+The ```node-agent-configs``` configMap may be used for other purpose of configuring node-agent in future, at present, there is only one kind of data ```data-path-concurrency```. Its data struct is like below: 
+
+```go
+type DataPathConcurrency struct {
```

Comment:
Not fully understand how this work. How about using the label selector instead? The key of the map is the label selector and the value is the current number.

Reply:
Here MatchLabel is the label selector. With the current approach, we want to give a shortcut of not specifying the selector all the time since `kubernetes.io/hostname` is used in most of the cases.

For example, we have below nodes in one cluster:

- Node-1 (kubernetes.io/hostname:cluster-node-1)
- Node-2 (kubernetes.io/hostname:cluster-node-2)
- Node-2 (kubernetes.io/hostname:cluster-node-3)

If I want to set an per-node number to Node-2, I make the struct like below:
```
Configs[cluster-node-2] = 3
```

In the other cluster, we don't want to use label `kubernetes.io/hostname` in the nodes, but another label `my-cluster/my-node-id`:
- Node-1 (my-cluster/my-node-id:00000)
- Node-2 (my-cluster/my-node-id:00001)
- Node-2 (my-cluster/my-node-id:00002)

Then I will have the struct like below:
```
MatchLabel = my-cluster/my-node-id
Configs[00001] = 3
```

If we use the label selector as the key directly, then users need to specify the per-node number as below:
```
Configs["kubernetes.io/hostname=cluster-node-2"] = 3
Configs["my-cluster/my-node-id=00001"] = 3
```

Finally, the current approach is simpler but the second approach (use the label selector as the key directly) looks more straight-forward.
So another proposal is like below:
If the key is with the label selector format (x=y), then we treat the key as a label selector
if the key is a single string (e.g., Configs[cluster-node-2] = 3), we treat it as a value of label ```kubernetes.io/hostname```

Let's vote on the 3 approaches.

Reply:
OK. Then let's use the straight way --- let's make it as `map[string]int `and have the keys like  "kubernetes.io/hostname=cluster-node-2" or "my-cluster/my-node-id=00001"

Reply:
Done on the changes according to the discussion.